axelor-message/src/main/java/com/axelor/apps/message/service/MessageServiceImpl.java

```diff
@@ -51,211 +53,254 @@
+            } else {
+                throw new AxelorException(message, IException.CONFIGURATION_ERROR, IExceptionMessage.MESSAGE_7);
+            }
+        }
+        if (replytoRecipients != null && !replytoRecipients.isEmpty()) {
+            mailBuilder.replyTo(Joiner.on(",").join(replytoRecipients));
+        }
+        if (toRecipients != null && !toRecipients.isEmpty()) {
+            mailBuilder.to(Joiner.on(",").join(toRecipients));
+        }
+        if (ccRecipients != null && !ccRecipients.isEmpty()) {
+            mailBuilder.cc(Joiner.on(",").join(ccRecipients));
+        }
+        if (bccRecipients != null && !bccRecipients.isEmpty()) {
+            mailBuilder.bcc(Joiner.on(",").join(bccRecipients));
+        }
+        if (!Strings.isNullOrEmpty(message.getContent())) {
+            mailBuilder.html(message.getContent());
+        }
+
+        for (MetaAttachment metaAttachment : getMetaAttachments(message)) {
+            MetaFile metaFile = metaAttachment.getMetaFile();
+            mailBuilder.attach(metaFile.getFileName(), MetaFiles.getPath(metaFile).toString());
+        }
+
+        mailBuilder.send();
+
+        message.setSentByEmail(true);
+        message.setStatusSelect(MessageRepository.STATUS_SENT);
+        message.setSentDateT(LocalDateTime.now());
+        message.setSenderUser(AuthUtils.getUser());
+
+        return messageRepository.save(message);
+
+    }
+
+    public Set<MetaAttachment> getMetaAttachments(Message message) {
+
+        Query<MetaAttachment> query = metaAttachmentRepository.all().filter("self.objectId = ?1 AND self.objectName = ?2", message.getId(), Message.class.getName());
+        return Sets.newHashSet(query.fetch());
+
+    }
+
+
+    public List<String> getEmailAddresses(Set<EmailAddress> emailAddressSet) {
+
+        List<String> recipients = Lists.newArrayList();
+        if (emailAddressSet != null) {
+            for (EmailAddress emailAddress : emailAddressSet) {
+
+                if (Strings.isNullOrEmpty(emailAddress.getAddress())) {
+                    continue;
+                }
+                recipients.add(emailAddress.getAddress());
+
+            }
+        }
+
+
+        return recipients;
+    }
+
+
+    @Override
+    public String printMessage(Message message) throws AxelorException {
+        return null;
+    }
+
+    @Override
+    @Transactional(rollbackOn = Exception.class)
+    public Message regenerateMessage(Message message) throws Exception {
+        Preconditions.checkNotNull(message.getRelatedTo1Select(), "cannot regenerate message without related model");
+        Class m = Class.forName(message.getRelatedTo1Select());
+        Model model = JPA.all(m).filter("self.id = ?", message.getRelatedTo1SelectId()).fetchOne();
+        Message newMessage = Beans.get(TemplateMessageService.class).generateMessage(model, message.getTemplate());
+        messageRepository.remove(message);
+        return newMessage;
+    }
```

Comment:
I think we have to add a M2O between Message and Model. 
Else, how know if the message is generated per a model or created manually by user ?

Reply:
Also I see you have already added tis M2O., so you have to use it here to get the model.